data-runtime/src/main/java/io/micronaut/data/runtime/operations/ExecutorReactiveOperationsSupportingCriteria.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2017-2024 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.data.runtime.operations;
+
+import io.micronaut.core.annotation.Experimental;
+import io.micronaut.data.operations.reactive.ReactiveCriteriaRepositoryOperations;
+import io.micronaut.data.runtime.convert.DataConversionService;
+import jakarta.persistence.criteria.CriteriaBuilder;
+import jakarta.persistence.criteria.CriteriaDelete;
+import jakarta.persistence.criteria.CriteriaQuery;
+import jakarta.persistence.criteria.CriteriaUpdate;
+import org.reactivestreams.Publisher;
+
+/**
+ * A variation of {@link ExecutorReactiveOperations} that supports {@link ReactiveCriteriaRepositoryOperations}.
+ * @author Denis Stepanov
+ */
+@Experimental
+public class ExecutorReactiveOperationsSupportingCriteria extends ExecutorReactiveOperations implements ReactiveCriteriaRepositoryOperations {
+
+    private final ExecutorAsyncOperationsSupportingCriteria asyncOperations;
+
+    public ExecutorReactiveOperationsSupportingCriteria(ExecutorAsyncOperationsSupportingCriteria asyncOperations,
+                                                        DataConversionService dataConversionService) {
+        super(asyncOperations, dataConversionService);
+        this.asyncOperations = asyncOperations;
+    }
+
+    @Override
+    public CriteriaBuilder getCriteriaBuilder() {
+        return asyncOperations.getCriteriaBuilder();
+    }
+
+    @Override
+    public <R> Publisher<R> findOne(CriteriaQuery<R> query) {
+        return fromCompletableFuture(() -> asyncOperations.findOne(query).toCompletableFuture());
+    }
+
+    @Override
+    public <T> Publisher<T> findAll(CriteriaQuery<T> query) {
+        return fromCompletableFuture(() -> asyncOperations.findAll(query).toCompletableFuture()).flatMapIterable(list -> list);
+    }
+
+    @Override
+    public <T> Publisher<T> findAll(CriteriaQuery<T> query, int offset, int limit) {
```

Comment:
I wonder why not accept the whole `Pageable` interface as an argument instead of specifying offset and limit?

Reply:
But we do want to be able to sort also no?

Reply:
I use the sorting in my case, and I do it via the Pageable interface. So if the sorting won't work after this PR, I will need to create another ticket :/

Reply:
Sorting is set into `CriteriaQuery`

Reply:
What are the difficulties of applying the sorting?

Reply:
> Sorting is set into CriteriaQuery

Not sure I understand

Reply:
See `CriteriaQuery#orderBy`

Reply:
So, what you are saying is, let's depend on the user using the `criteriaQuery.orderBy` instead of the `Pageable` interface. Is that true?

Reply:
And btw, the `buildQuery` method only actually does something in the sole case of a `FIND_ALL`. I think we should change that to include any `FIND_PAGE` and `FIND_ONE` as well.

Reply:
Check: https://github.com/micronaut-projects/micronaut-data/blob/4df20033f12137b272101d13a6940b2c5172abcd/data-runtime/src/main/java/io/micronaut/data/runtime/intercept/criteria/AbstractSpecificationInterceptor.java#L338

Reply:
Why find one?

Reply:
The `findOne` basically sets a `limit 1` right? But maybe we want the values to be sorted before applying `limit 1`